### PR TITLE
ML-025/add user tests

### PR DESCRIPTION
**Pull request by twill11**
https://trello.com/c/ZZRvgwNZ/7-ml-025-add-user-tests

### CODE REVIEW

**Comment by dmitrymalllpr** [comment]
Please fix issues and change label

File: db/migrate/20220220190037_devise_token_auth_create_users.rb [hunk 1]
```diff
@@ -53,5 +53,6 @@ def change
     add_index :users, :reset_password_token, unique: true
     add_index :users, :confirmation_token,   unique: true
     add_index :users, :unlock_token,         unique: true
+    add_index :users, :username,             unique: true
```
Comment (WhatWasThatt) [inline db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: we have decision to make validations on model level

Comment (WhatWasThatt) [inline db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: we have decision to make validations on model level

**Review by WhatWasThatt — CHANGES_REQUESTED** [review]

File: db/migrate/20220220190037_devise_token_auth_create_users.rb (hunk 1 above)
Comment (twill11) [reply db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: I don't quite understand, can you rephrase?

Comment (WhatWasThatt) [reply db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: as I understand the issue, we need to leave validations on the model level as you did before the last commit. but I'm not sure and ask help

Comment (twill11) [reply db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: ok, thanks for your review!

**Review by WhatWasThatt — APPROVED** [review]

File: app/models/user.rb [hunk 2]
```diff
@@ -8,4 +8,9 @@ class User < ActiveRecord::Base
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable,
          :lockable, :confirmable, :trackable
+
+  validates :name, length: { in: 2..25 }
+  validates :surname, length: { in: 2..25 }
+  validates :username, presence: true, length: { in: 4..20 }
```
Comment (WhatWasThatt) [inline app/models/user.rb:14]: we need to validate the username

**Review by WhatWasThatt — CHANGES_REQUESTED** [review]

**Review by WhatWasThatt — APPROVED** [review]
Great job!

**Review by WhatWasThatt — DISMISSED** [review]

**Comment by KarinaKlevzhits** [comment]
Please add link to task in trello

File: db/migrate/20220220190037_devise_token_auth_create_users.rb (hunk 1 above)
Comment (ShrewHools) [inline db/migrate/20220220190037_devise_token_auth_create_users.rb:56]: should everyone roll back this migration and run it again?
or are there other solutions?

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  default_url_options :host => 'localhost:3000' # this code-line is for request tests
```
Comment (ShrewHools) [inline config/routes.rb:2]: is it a good idea to write a specific thing in the routes.rb file for tests or for dev?
Or maybe there are other solutions?
are there any files that change the config just for testing purposes?

File: db/seeds.rb [hunk 4]
```diff
@@ -10,3 +10,4 @@
 Role.create(name: 'redactor', permission_attributes: { can_lock_user: false, can_edit_role: false, can_set_role: false, can_read: true, can_edit: true })
 Role.create(name: 'reviewer', permission_attributes: { can_lock_user: false, can_edit_role: false, can_set_role: false, can_read: true, can_edit: false })
 FactoryBot.create_list(:tag, 20)
+FactoryBot.create_list(:user, 5)
```
Comment (ShrewHools) [inline db/seeds.rb:13]: maybe you can create users with different roles, known to us email and passwords?
Wouldn't it be nice and convenient for team?

**Comment by ShrewHools** [comment]
and please merge updated master into your branch, thank you)

File: config/routes.rb (hunk 3 above)
Comment (twill11) [reply config/routes.rb:2]: i have problems with request spec for devise token auth, and it's only one solution which helped me, i try to improve this for configuration file

**Comment by dmitrymalllpr** [comment]
devise_for ???

File: config/routes.rb (hunk 3 above)
Comment (ShrewHools) [reply config/routes.rb:2]: config/environments/test.rb file?)

File: db/seeds.rb [hunk 5]
```diff
@@ -6,7 +6,31 @@
...
+                        { can_lock_user: true, can_edit_role: true,
+                          can_set_role: true, can_read: true, can_edit: true })
+redactor_role = Role.create(name: 'redactor', permission_attributes:
+                        { can_lock_user: false, can_edit_role: false,
+                          can_set_role: false, can_read: true, can_edit: true })
+reviewer_role = Role.create(name: 'reviewer', permission_attributes:
+                        { can_lock_user: false, can_edit_role: false,
+                          can_set_role: false, can_read: true, can_edit: false })
+
+admin = User.new(email: 'admin@example.com', password: '123456', password_confirmation: '123456',
```
Comment (ShrewHools) [inline db/seeds.rb:19]: User.create(..., confirmed_at: Time.now) instead of 2 lines below. Or am I missing something and skip_confirmation! necessary?

File: config/database.yml [hunk 6]
```diff
@@ -57,7 +57,7 @@ development:
 # Do not set this db to the same as development or production.
 test:
   <<: *default
-  host: localhost
+  host: ''
```
Comment (ShrewHools) [inline config/database.yml:60]: ??

File: db/seeds.rb [hunk 7]
```diff
@@ -6,7 +6,26 @@
...
+                        { can_lock_user: true, can_edit_role: true,
+                          can_set_role: true, can_read: true, can_edit: true })
+redactor_role = Role.create(name: 'redactor', permission_attributes:
+                        { can_lock_user: false, can_edit_role: false,
+                          can_set_role: false, can_read: true, can_edit: true })
+reviewer_role = Role.create(name: 'reviewer', permission_attributes:
+                        { can_lock_user: false, can_edit_role: false,
+                          can_set_role: false, can_read: true, can_edit: false })
+
+User.create(email: 'admin@example.com', password: '123456', password_confirmation: '123456',
```
Comment (WhatWasThatt) [inline db/seeds.rb:19]: you can use traits

**Review by WhatWasThatt — APPROVED** [review]

**Review by ShrewHools — APPROVED** [review]

**Comment by ShrewHools** [comment]
gj
